redis/src/cluster_client.rs

```diff
@@ -41,69 +43,72 @@ impl ClusterClientBuilder {
     /// Upon failure to parse initial nodes or if the initial nodes have different passwords or
     /// usernames, an error is returned.
     pub fn build(self) -> RedisResult<ClusterClient> {
-        let initial_nodes = self.initial_nodes?;
+        let mut initial_nodes = self.initial_nodes?;
 
-        let mut nodes = Vec::with_capacity(initial_nodes.len());
-        let mut connection_info_password = None::<String>;
-        let mut connection_info_username = None::<String>;
+        let mut cluster_params = self.cluster_params;
+        let password = if cluster_params.password.is_none() {
+            cluster_params.password = initial_nodes[0].redis.password.clone();
+            &cluster_params.password
+        } else {
+            &None
+        };
+        let username = if cluster_params.username.is_none() {
+            cluster_params.username = initial_nodes[0].redis.username.clone();
+            &cluster_params.username
+        } else {
+            &None
+        };
 
-        for (index, info) in initial_nodes.into_iter().enumerate() {
-            if let ConnectionAddr::Unix(_) = info.addr {
-                return Err(RedisError::from((ErrorKind::InvalidClientConfig,
-                                         "This library cannot use unix socket because Redis's cluster command returns only cluster's IP and port.")));
-            }
+        initial_nodes = initial_nodes
```

Comment:
What's the impetus here for moving from a `for`-loop to iterator combinators? IMO in this case it doesn't much improve the code (and this is another change that is unrelated to the stated goal of the commit) -- in fact, it looks like it adds another level of indentation which makes the code harder to read.

Reply:
I initially thought that this function is called each time we need a new connection, so I changed it to an iterator pattern which would be (slightly) faster.

Reply:
How/why do you think it would be faster? I don't think that's true.

Reply:
I'm relatively new to rust, but from what I've read, iterators can be parallelised better than for-loops & also we don't have to check & increase the capacity of the vector that we collect into. It is a very slight improvement.

Ref: https://users.rust-lang.org/t/we-all-know-iter-is-faster-than-loop-but-why/51486

Reply:
Okay, please don't resolve my review comments unless you've done as I suggested.

There will be no meaningful difference in performance here, especially if you allocate the new `Vec` by using `with_capacity()` up front. The `for`-loop is more readable in this case.